eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs

```diff
@@ -0,0 +1,285 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Ledger.Conway.Rules.Ledger (
+  ConwayLEDGER,
+  ConwayLedgerPredFailure (..),
+) where
+
+import Cardano.Crypto.DSIGN (DSIGNAlgorithm (..))
+import Cardano.Crypto.Hash.Class (Hash)
+import Cardano.Ledger.Alonzo.Rules (AlonzoUtxowEvent)
+import Cardano.Ledger.Alonzo.Scripts (AlonzoScript, CostModels (..))
+import Cardano.Ledger.Alonzo.UTxO (AlonzoScriptsNeeded (..))
+import Cardano.Ledger.Babbage.Core (BabbageEraTxBody)
+import Cardano.Ledger.Babbage.Rules (BabbageUTXOW, BabbageUtxowPredFailure)
+import Cardano.Ledger.Babbage.Tx (IsValid (..))
+import Cardano.Ledger.Babbage.TxBody (BabbageTxOut (..))
+import Cardano.Ledger.BaseTypes (ShelleyBase)
+import Cardano.Ledger.Block (txid)
+import Cardano.Ledger.Conway.Core (ConwayEraTxBody (..))
+import Cardano.Ledger.Conway.Era (ConwayLEDGER, ConwayTALLY)
+import Cardano.Ledger.Conway.Governance (ConwayTallyState)
+import Cardano.Ledger.Conway.Rules.Tally (GovernanceProcedure (..), TallyEnv (..), ConwayTallyPredFailure)
+import Cardano.Ledger.Conway.Tx (AlonzoEraTx (..))
+import Cardano.Ledger.Core (Era (..), EraIndependentTxBody, EraPParams (..), EraRule, EraScript (..), EraTx (..), EraTxOut (..))
+import Cardano.Ledger.Crypto (Crypto (..))
+import Cardano.Ledger.Shelley.Core (EraTallyState (..), ShelleyEraTxBody (..))
+import Cardano.Ledger.Shelley.Delegation.Certificates (DCert)
+import Cardano.Ledger.Shelley.LedgerState (DPState (..), DState (..), LedgerState (..), PPUPState, UTxOState (..), obligationDPState)
+import Cardano.Ledger.Shelley.Rules (
+  DelegsEnv (..),
+  DelplEnv (..),
+  LedgerEnv (..),
+  ShelleyDELEGS,
+  ShelleyDelegsEvent,
+  ShelleyDelegsPredFailure,
+  ShelleyLEDGERS,
+  ShelleyLedgersEvent (..),
+  ShelleyLedgersPredFailure (..),
+  UtxoEnv (..),
+ )
+import Cardano.Ledger.UTxO (EraUTxO (..))
+import Control.State.Transition.Extended (
+  Assertion (..),
+  AssertionViolation (..),
+  Embed (..),
+  STS (..),
+  TRC (..),
+  TransitionRule,
+  judgmentContext,
+  trans,
+ )
+import Data.Kind (Type)
+import Data.Sequence (Seq)
+import qualified Data.Sequence.Strict as StrictSeq
+import GHC.Generics (Generic (..))
+import GHC.Records (HasField)
+import Lens.Micro ((^.))
+
+data ConwayLedgerPredFailure era
+  = UtxowFailure (PredicateFailure (EraRule "UTXOW" era)) -- Subtransition Failures
+  | DelegsFailure (PredicateFailure (EraRule "DELEGS" era)) -- Subtransition Failures
+  | TallyFailure (PredicateFailure (EraRule "TALLY" era)) -- Subtransition Failures
```

Comment:
> inject the Shelley failures inside the Conway ledger failures with a ShelleyLedgerFailure (ShelleyLedgerPredFailure era)

But you aren't invoking the LEDGER rule from Shelley, so there is no reason to wrap its predicate failure. This will only lead to confusion. Also, the fact that you can save one line of code isn't a good motivation IMHO

Reply:
We're already doing it for example here:
```haskell
data BabbageUtxoPredFailure era
  = AlonzoInBabbageUtxoPredFailure !(AlonzoUtxoPredFailure era) -- Inherited from Alonzo
  | ...
```

I don't see why we can't do it in Conway, since the Conway LEDGER rule still calls the same rules as the Shelley, plus the TALLY rule.

Reply:
That's because we are calling these validation rules from alonzo, which themselves produce `AlonzoUtxoPredFailure`:
https://github.com/input-output-hk/cardano-ledger/blob/24d7a68b00c2e71c4ba9491bacb6be1970015602/eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs#L34-L41

In your PR however, you are not calling any piece of code that produces `ShelleyLedgerPredFailure`, and instead you would be constructing this Shelley failure in Conway yourself. Does this clarify my reasoning a bit more?

Reply:
>  Conway LEDGER rule still calls the same rules as the Shelley, plus the TALLY rule.

But it doesn't call the `ShelleyLEDGER` rule, which that predicate failure is attached to

Reply:
Ah, I see, that makes sense now. I'll prepend `Conway` and then maybe create an issue to rename the `ShelleyLedgerPredFailure` constructors as well?

Reply:
Awesome!!!